source/fe/mapping_q_cache.cc

```diff
@@ -0,0 +1,121 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/work_stream.h>
+
+#include <deal.II/fe/mapping_q_cache.h>
+
+#include <functional>
+
+DEAL_II_NAMESPACE_OPEN
+
+template <int dim, int spacedim>
+MappingQCache<dim, spacedim>::MappingQCache(
+  const unsigned int polynomial_degree)
+  : MappingQGeneric<dim, spacedim>(polynomial_degree)
+{}
+
+
+
+template <int dim, int spacedim>
+MappingQCache<dim, spacedim>::MappingQCache(
+  const MappingQCache<dim, spacedim> &mapping)
+  : MappingQGeneric<dim, spacedim>(mapping)
+  , support_point_cache(mapping.support_point_cache)
+{}
+
+
+
+template <int dim, int spacedim>
+MappingQCache<dim, spacedim>::~MappingQCache()
+{
+  if (clear_signal.connected())
+    clear_signal.disconnect();
+}
+
+
+
+template <int dim, int spacedim>
+std::unique_ptr<Mapping<dim, spacedim>>
+MappingQCache<dim, spacedim>::clone() const
+{
+  return std_cxx14::make_unique<MappingQCache<dim, spacedim>>(*this);
+}
+
+
+
+template <int dim, int spacedim>
+bool
+MappingQCache<dim, spacedim>::preserves_vertex_locations() const
+{
+  return false;
+}
+
+
+
+template <int dim, int spacedim>
+void
+MappingQCache<dim, spacedim>::initialize(
+  const Triangulation<dim, spacedim> &  triangulation,
+  const MappingQGeneric<dim, spacedim> &mapping)
+{
+  AssertDimension(this->get_degree(), mapping.get_degree());
+
+  clear_signal = triangulation.signals.any_change.connect(
+    [&]() -> void { this->support_point_cache.reset(); });
+
+  support_point_cache =
+    std::make_shared<std::vector<std::vector<std::vector<Point<spacedim>>>>>(
+      triangulation.n_levels());
+  for (unsigned int l = 0; l < triangulation.n_levels(); ++l)
+    (*support_point_cache)[l].resize(triangulation.n_raw_cells(l));
+
+  WorkStream::run(
+    triangulation.begin(),
+    triangulation.end(),
```

Comment:
Do we need this for non-active cells? What about cells that are artificial?

Reply:
Yes, we need also non-active cells because we want to be able to use it on the multigrid levels as well. We do not need the artificial cells, but I did not try to adjust that because the logic to get this right for both the active cells and the level cells seemed not worth the effort. I can try that if you insist, but artificial cells are a limited fraction of all cells according to our experiments from 2010.